### DISCUSSION TITLE
원시타입과 참조타입의 차이점은 무엇인가요?

### DISCUSSION THREAD

**Discussion by pipisebastian** [publdaze/js-study#44] [원시 값과 객체의 비교]
- 원시타입과 참조타입의 차이점은 무엇인가요?
  - const로 선언한 객체를 수정할 수 있는 이유는 무엇인가요?
  - 불변성을 지니는 원시타입의 장점은 무엇인가요? 
    - 상수와 불변의 차이는 무엇일까요?
  - 객체를 불변하게 만들기 위해 어떤 방법을 사용할 수 있나요?
  - 원시 값과 객체는 메모리 영역 중 각각 어디에 저장이 되나요? 해당 메모리 영역을 사용하는 이유도 말해주세요.

**Comment by pipisebastian** [comment]
> 원시타입과 참조타입의 차이점은 무엇인가요? 

- **원시 타입**은 변경 불가능(읽기 전용)하며, 변수에 값이 직접 저장됩니다. `값에 의한 전달`
    - 그렇기 때문에 변수에 값을 재할당한다는 것은, 값을 직접 바꾸는게 아니라, 새로운 메모리 공간을 확보하고, 변수가 새로운 메모리로 교체하게 됩니다.
- **참조 타입**은 변경 가능하며, 변수에는 해당 객체의 메모리 주소가 저장됩니다. `참조에 의한 전달`
    - 참조 타입은 어떤 타입의 값이 들어갈지 모르고, 가변적인 크기 때문에, 미리 여유롭게 메모리를 확보합니다.
   
> const로 선언한 객체를 수정할 수 있는 이유는 무엇인가요?

- 객체는 기본적으로 원시타입과 다르게 값이 아닌, 메모리주소를 저장합니다.
- 이 메모리 주소는 또 다시 객체가 담긴 값들의 주소를 저장하는데요, 그렇기 때문에 객체의 값을 수정한다는 것은, 실제로는 한단계를 더 거쳐서 값이 변하는 거라, 1차적인 변수를 저장하는 메모리 주소는 변하지 않습니다.

> 불변성을 지니는 원시타입의 장점은 무엇인가요?

- 원시값은 항상 변하지 않으므로, 데이터의 신뢰성을 보장합니다.

> 상수와 불변의 차이는 무엇일까요?

- 불변이란, 메모리에 들어있는 값을 변경하지 못합니다. 원시타입의 경우 불변성을 지니기에, 메모리의 값을 직접 수정하지 못하고, 값을 재할당하기 위해서는 새로운 메모리를 할당받아야합니다.
- 상수란, 값을 재할당하지 못하는 변수입니다.
- 그렇기 때문에 불변이란 메모리영역 관점에서 메모리 값을 못 바꾼다는 것이고, 상수는 재할당이 불가능한 것입니다.

> 객체를 불변하게 만들기 위해 어떤 방법을 사용할 수 있나요?

- `object.freeze`
    - 해당 방법은 객체의 중첩된 객체까지는 동결시키지 못합니다. 즉, 한단계까지만 동결을 시키는데요.
    - 중첩된 객체까지 동결하기 위해서는 재귀적으로 `object.freeze` 를 사용해야합니다.
    - 그러나, 해당 방법은 객체의 프로퍼티를 변경할 수는 없지만, 만약 let, var 로 선언할 경우, 객체의 재할당은 가능합니다.
    - 그렇기 때문에 `const`로 선언하여 재할당까지 막아준다면, 객체의 재할당과 내부 속성 변경 모두를 막을 수 있습니다.
    
> 원시 값과 객체는 메모리 영역 중 각각 어디에 저장이 되나요? 해당 메모리 영역을 사용하는 이유도 말해주세요.

- 원시타입은 메모리 스택영역에 저장되고, 참조타입은 힙 영역에 저장됩니다.
- 참조타입(객체)의 경우, 동적으로 크기가 변할 수 있기 때문에 힙 영역에 저장하며, 원시타입의 경우 불변성을 지니고, 크기가 고정되어있어 스택영역에 할당된다.

**Comment by publdaze** [comment]
> - 원시타입과 참조타입의 차이점은 무엇인가요?
- 원시타입은 불변이며, 변수에 값이 할당됨.
- 참조타입은 변경 가능, 변수에 메모리 주소가 할당됨.
> - const로 선언한 객체를 수정할 수 있는 이유는 무엇인가요?
- const로 선언된 변수의 값은 재할당이 불가능한데, 객체의 프로퍼티를 수정할 때는 변수의 값인 메모리 주소가 변하는 게 아닌 메모리 주소가 참조하고 있는 메모리 공간의 값이 변하는 것이기 때문에 수정 가능
> - 불변성을 지니는 원시타입의 장점은 무엇인가요? 
- 불변성이 보장되면, 의도치 않게 변경되지 않기 때문에 값이 예측이 가능하고 부수 효과 없이 안전하게 데이터를 다룰 수 있음.
> - 상수와 불변의 차이는 무엇일까요?
- 상수는 재할당이 불가능한 변수이고, 불변은 데이터 자체의 값이 변경되지 않는 것.
> - 객체를 불변하게 만들기 위해 어떤 방법을 사용할 수 있나요?
- Object.freeze 메소드를 사용하거나 Object.seal 메소드를 사용하더나 별도의 라이브러리를 사용하는 방법이 있음
> - 원시 값과 객체는 메모리 영역 중 각각 어디에 저장이 되나요? 해당 메모리 영역을 사용하는 이유도 말해주세요.
- 원시 값은 값이 변하지 않고 정적이기 때문에 스택에 저장되고, 객체는 변하는 값이라 동적으로 변할 수 있기 때문에 런타임에 메모리 공간의 크기를 결정해야 하므로 힙에 저장된다.
